mempool/clist_sidecar.go

```diff
@@ -0,0 +1,655 @@
+package mempool
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/clist"
+	"github.com/tendermint/tendermint/libs/log"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	"github.com/tendermint/tendermint/mev"
+	"github.com/tendermint/tendermint/types"
+)
+
+// CListPriorityTxSidecar is an ordered in-memory pool for transactions stored
+// on sentry nodes that validators can pull from. Transaction validity is not checked using the
+// CheckTx abci message since no transactions are added to the mempool. The
+// CListPriorityTxSidecar uses a concurrent list structure for storing transactions that can
+// be efficiently accessed by multiple concurrent readers.
+type CListPriorityTxSidecar struct {
+	// Atomic integers
+	height                 int64 // the last block Update()'d to
+	heightForFiringAuction int64 // the height of the block to fire the auction for
+	txsBytes               int64 // total size of sidecar, in bytes
+	lastBundleHeight       int64 // height of last accepted bundle tx, for status rpc purposes
+
+	// notify listeners (ie. consensus) when txs are available
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // fires once for each height, when the mempool is not empty
+
+	txs    *clist.CList // concurrent linked-list of good SidecarTxs
+	txsMap sync.Map
+
+	// sync.Map: Key{height, bundleID} -> Bundle{
+	// // height int64
+	// // enforcedSize int64
+	// // currSize int64
+	// // bundleID int64
+	// // sync.Map bundleOrder -> *SidecarTx
+	// }
+	bundles     sync.Map
+	maxBundleID int64
+
+	updateMtx tmsync.RWMutex
```

Comment:
I am not sure I understand how this `RWMutex` is used. As far as I can see, it is used in `AddTx`, `ReapMaxTxs`, `Lock` and `Unlock`. `AddTx` and `ReapMaxTxs` grab a read lock and `Lock` and `Unlock` grabs a read-write lock. Since `Lock` is never used by a caller, what's the point of the rwmutex?

Reply:
We lock the sidecar on Commit now, as suggested in https://github.com/informalsystems/audit-skip/issues/7, so this should now be useful.